File: examples/java/pact_validation_with_oas/one_consumer_one_provider/provider/src/test/java/br/com/zup/pact/provider/service/AccountServiceImplTest.java
package br.com.zup.pact.provider.service;

import br.com.zup.pact.provider.dto.AccountDetailsDTO;
import br.com.zup.pact.provider.dto.BalanceDTO;
import br.com.zup.pact.provider.entity.Account;
import br.com.zup.pact.provider.exception.ClientNotFoundException;
import br.com.zup.pact.provider.repository.AccountRepository;
import br.com.zup.pact.provider.service.impl.AccountServiceImpl;
import br.com.zup.pact.provider.stub.AccountStub;
import org.junit.Assert;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.mockito.internal.matchers.apachecommons.ReflectionEquals;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.when;

@SpringBootTest
public class AccountServiceImplTest {

    @MockBean
    private AccountRepository accountRepository;

    @Autowired
    private AccountService accountService;

    @Autowired
    private AccountStub accountStub;

    @Test
    void testGetAccountDetailsByAccountIdValid() {
        final AccountDetailsDTO accountDetailsDTOMock = Account.fromEntityToDto(accountStub.getAccounts().get(1));
        when(accountRepository.findByAccountId(anyInt())).thenReturn(Optional.of(accountDetailsDTOMock));

        final AccountDetailsDTO accountDetailsDTOActual = accountService.getAccountDetailsByAccountId(anyInt()).orElse(null);

        Assert.assertTrue(new ReflectionEquals(accountDetailsDTOMock).matches(accountDetailsDTOActual));

    }

    @Test
    void testGetAccountDetailsByAccountIdInvalid() {
        when(accountRepository.findByAccountId(anyInt()))
                    .thenThrow(ClientNotFoundException.class);

        Assertions.assertThrows(ClientNotFoundException.class, () -> accountService.getAccountDetailsByAccountId(anyInt()));

    }

    @Test
    void testGetAllAccounts() {
        final List<AccountDetailsDTO> allAccountsDTOMock = Account.fromEntityToDtoList(
                new ArrayList<>(accountStub.getAccounts().values())
        );
        when(accountRepository.getAll()).thenReturn(Optional.of(allAccountsDTOMock));
        final List<AccountDetailsDTO> allAccountsDTOActual = accountService.getAll().orElse(null);

        Assert.assertTrue(new ReflectionEquals(allAccountsDTOActual).matches(allAccountsDTOMock));
    }

    @Test
    void testGetAllAccountsEmpty() {
        when(accountRepository.getAll())
                .thenReturn(Optional.empty());

        final Optional<List<AccountDetailsDTO>> all = accountService.getAll();
        assertThat(all.isEmpty());
    }


    @Test
    void testGetBalanceByAccountId() {
        final BalanceDTO balanceDTOMock = BalanceDTO.fromAccountToDTO(
                accountStub.getAccounts().get(1)
        );
        when(accountRepository.getBalanceByAccountId(anyInt())).thenReturn(Optional.of(balanceDTOMock));

        final BalanceDTO BalanceDTOActual = accountService.getBalanceByAccountId(anyInt()).orElse(null);

        Assert.assertTrue(new ReflectionEquals(balanceDTOMock).matches(BalanceDTOActual));
    }

    @Test
    void testGetBalanceByAccountIdEmpty() {
        when(accountRepository.getBalanceByAccountId(anyInt()))
                .thenReturn(Optional.empty());

        final Optional<BalanceDTO> balanceDTO = accountService.getBalanceByAccountId(anyInt());

        assertThat(balanceDTO.isEmpty());
    }

}

File: example/java/kotlin/provider/src/test/kotlin/br/com/zup/pact/provider/service/AccountServiceImplTest.kt
package br.com.zup.pact.provider.service

import br.com.zup.pact.provider.dto.AccountDetailsDTO
import br.com.zup.pact.provider.dto.BalanceDTO
import br.com.zup.pact.provider.enums.AccountType
import br.com.zup.pact.provider.repository.AccountRepository
import io.mockk.clearMocks
import io.mockk.every
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.util.*

@ExtendWith(MockKExtension::class)
class AccountServiceImplTest {

    private val accountRepositoryMock: AccountRepository = mockk()
    private val accountService: AccountService = AccountServiceImpl(accountRepositoryMock)

    @BeforeEach
    fun setUp() {
        clearMocks(accountRepositoryMock)
    }

    @Test
    fun `Method getAll should return getAll from the repository`() {

        every { accountRepositoryMock.getAll() }
                .returns(listOf(AccountDetailsDTO(
                        accountId = 1,
                        balance = 100.0,
                        accountType = AccountType.MASTER
                )))

        val expectedAccountDetailsDTO = accountService.getAll()

        Assertions.assertThat(expectedAccountDetailsDTO).isEqualTo(
                listOf(AccountDetailsDTO(
                        accountId = 1,
                        balance = 100.0,
                        accountType = AccountType.MASTER
                ))
        )
    }

    @Test
    fun `Method findByAccountId should return findByAccountId from the repository`() {

        every { accountRepositoryMock.findByAccountId(1) }
                .returns(Optional.of(BalanceDTO(
                        accountId = 1,
                        balance = 100.0
                )))

        val expectedBalanceDTO = accountService.findByAccountId(1)

        Assertions.assertThat(expectedBalanceDTO).isEqualTo(
                Optional.of(BalanceDTO(
                        accountId = 1,
                        balance = 100.0
                ))
        )
    }
}